conn.go

```diff
@@ -248,170 +301,121 @@ func (c *Conn) startup(cfg *ConnConfig) error {
-	return c.r.Read(p)
+	if c.started {
+		c.pool.HandleError(c, err, true)
+	}
 }
 
-func (c *Conn) recv() (frame, error) {
-	size := headerProtoSize[c.version]
-	resp := make(frame, size, size+512)
+func (c *Conn) recv() error {
+	// not safe for concurrent reads
 
 	// read a full header, ignore timeouts, as this is being ran in a loop
-	c.conn.SetReadDeadline(time.Time{})
-	_, err := io.ReadFull(c.r, resp[:size])
-	if err != nil {
-		return nil, err
+	// TODO: TCP level deadlines? or just query level deadlines?
+	if c.timeout > 0 {
+		c.conn.SetReadDeadline(time.Time{})
 	}
 
-	if v := c.version | flagResponse; resp[0] != v {
-		return nil, NewErrProtocol("recv: response protocol version does not match connection protocol version (%d != %d)", resp[0], v)
+	// were just reading headers over and over and copy bodies
+	head, err := readHeader(c.r, c.headerBuf)
+	if err != nil {
+		return err
 	}
 
-	bodySize := resp.Length(c.version)
-	if bodySize == 0 {
-		return resp, nil
+	call := &c.calls[head.stream]
+	call.mu.Lock()
+	err = call.framer.readFrame(&head)
+	call.mu.Unlock()
+	if err != nil {
+		return err
 	}
-	resp.grow(bodySize)
-
-	const maxAttempts = 5
-
-	n := size
-	for i := 0; i < maxAttempts; i++ {
-		var nn int
-		c.conn.SetReadDeadline(time.Now().Add(c.timeout))
-		nn, err = io.ReadFull(c.r, resp[n:size+bodySize])
-		if err == nil {
-			break
-		}
-		n += nn
 
-		if verr, ok := err.(net.Error); !ok || !verr.Temporary() {
-			break
-		}
-	}
+	// once we get to here we know that the caller must be waiting and that there
+	// is no error.
+	call.resp <- nil
+	c.uniq <- head.stream
 
-	if err != nil {
-		return nil, err
-	}
+	return nil
+}
 
-	return resp, nil
+type callReq struct {
+	// could use a waitgroup but this allows us to do timeouts on the read/send
+	resp   chan error
+	mu     sync.Mutex
+	framer *framer
 }
 
-func (c *Conn) execSimple(op operation) (interface{}, error) {
-	f, err := op.encodeFrame(c.version, nil)
-	if err != nil {
-		// this should be a noop err
-		return nil, err
-	}
+func (c *Conn) exec(req frameWriter, tracer Tracer) (frame, error) {
+	// TODO: move tracer onto conn
+	stream := <-c.uniq
 
-	bodyLen := len(f) - headerProtoSize[c.version]
-	f.setLength(bodyLen, c.version)
+	call := &c.calls[stream]
+	// resp is basically a waiting semaphore protecting the framer
+	framer := newFramer(c, c, c.compressor, c.version)
+	defer framerPool.Put(framer)
 
-	if _, err := c.Write([]byte(f)); err != nil {
-		c.Close()
-		return nil, err
+	call.framer = framer
+
+	if tracer != nil {
+		framer.trace()
 	}
 
-	// here recv wont timeout waiting for a header, should it?
-	if f, err = c.recv(); err != nil {
+	// there is a race that we can read and write to the same buffer, I dont think
+	// the data will actually corrupt but to be safe and appease the race detector gods,
+	// guard it.
+	// We could fix this by using seperate read and write buffers, which may end up
+	// being faster and easier to reason about.
```

Comment:
Not really, just the change on read and write paths, currently they just share the same buffer and assume the owner of framer is not reading/writing concurrently, as it should be used for a single stream on a connection.